packages/middleware-compression/src/compressionMiddleware.ts

```diff
@@ -0,0 +1,102 @@
+import { HttpRequest } from "@smithy/protocol-http";
+import {
+  AbsoluteLocation,
+  BuildHandler,
+  BuildHandlerArguments,
+  BuildHandlerOptions,
+  BuildHandlerOutput,
+  BuildMiddleware,
+  MetadataBearer,
+} from "@smithy/types";
+
+import { compressStream } from "./compressStream";
+import { compressString } from "./compressString";
+import { CompressionResolvedConfig } from "./configurations";
+import { CLIENT_SUPPORTED_ALGORITHMS, CompressionAlgorithm } from "./constants";
+import { isStreaming } from "./isStreaming";
+
+/**
+ * @internal
+ */
+export interface CompressionMiddlewareConfig {
+  /**
+   * Defines the priority-ordered list of compression algorithms supported by the service operation.
+   */
+  encodings: string[];
+
+  /**
+   * Indicates that the streaming blob MUST be finite and have a known size when sending data from a client to a server.
+   * Populated if smithy requiresLength is set https://smithy.io/2.0/spec/streaming.html#requireslength-trait
+   */
+  streamRequiresLength?: boolean;
+}
+
+/**
+ * @internal
+ */
+export const compressionMiddleware =
+  (config: CompressionResolvedConfig, middlewareConfig: CompressionMiddlewareConfig): BuildMiddleware<any, any> =>
+  <Output extends MetadataBearer>(next: BuildHandler<any, Output>): BuildHandler<any, Output> =>
+  async (args: BuildHandlerArguments<any>): Promise<BuildHandlerOutput<Output>> => {
+    if (!HttpRequest.isInstance(args.request) || config.disableRequestCompression) {
+      return next(args);
+    }
+
+    const { request } = args;
+    const { body, headers } = request;
+    const { encodings, streamRequiresLength } = middlewareConfig;
+
+    let updatedBody = body;
+    let updatedHeaders = headers;
+
+    for (const algorithm of encodings) {
+      if (CLIENT_SUPPORTED_ALGORITHMS.includes(algorithm as CompressionAlgorithm)) {
+        let isRequestCompressed = false;
+        if (isStreaming(body)) {
+          if (!streamRequiresLength) {
+            updatedBody = await compressStream(body);
+            isRequestCompressed = true;
+          } else {
+            // Invalid case. We should never get here.
+            throw new Error("Compression is not supported for streaming blobs that require a length.");
+          }
+        } else if (body.length >= config.requestMinCompressionSizeBytes) {
```

Comment:
nit: is body.length okay here or would using [calculateBodyLength](https://github.com/smithy-lang/smithy-typescript/blob/cb044e77827c9f97b75167be61b6d39beb03ad1e/packages/util-body-length-node/src/calculateBodyLength.ts#L6) be a little more robust?

Reply:
Good catch! Fixed in [`f2ba37b` (#5620)](https://github.com/aws/aws-sdk-js-v3/pull/5620/commits/f2ba37b8859c4bfdc66c71e84d32d2daedc096ce)